File: php/src/Core/Sse.php
<?php

declare(strict_types=1);

namespace MisarMail\Core;

use MisarMail\ApiError;

/**
 * Server-Sent Events client for the MisarMail streaming endpoints.
 *
 * Both streams frame events as "data: <json>" and close with the sentinel
 * "data: [DONE]". One of the two is a POST, so this uses a cURL write callback
 * rather than an EventSource-style helper.
 */
final class Sse
{
    private const DONE = '[DONE]';

    /**
     * Opens an SSE endpoint and hands each decoded frame to $onEvent until the
     * stream terminates. Blocks for the life of the stream.
     *
     * @param callable(array<string, mixed>): void $onEvent
     */
    public static function stream(
        string $url,
        string $apiKey,
        string $method,
        mixed $body,
        callable $onEvent,
    ): void {
        $handle = curl_init($url);
        $headers = [
            'Authorization: Bearer ' . $apiKey,
            'Accept: text/event-stream',
        ];

        if ($body !== null) {
            $headers[] = 'Content-Type: application/json';
        }

        $buffer = '';
        $stopped = false;

        curl_setopt_array($handle, [
            CURLOPT_CUSTOMREQUEST => strtoupper($method),
            CURLOPT_HTTPHEADER => $headers,
            // A stream has no useful deadline.
            CURLOPT_TIMEOUT => 0,
            CURLOPT_WRITEFUNCTION => function ($_handle, string $chunk) use (
                &$buffer, &$stopped, $onEvent
            ): int {
                $length = strlen($chunk);
                if ($stopped) {
                    return $length;
                }

                $buffer .= $chunk;

                while (($index = strpos($buffer, "\n")) !== false) {
                    $line = rtrim(substr($buffer, 0, $index), "\r");
                    $buffer = substr($buffer, $index + 1);

                    if (!str_starts_with($line, 'data:')) {
                        continue;
                    }

                    $payload = trim(substr($line, 5));
                    if ($payload === self::DONE) {
                        $stopped = true;
                        return $length;
                    }
                    if ($payload === '') {
                        continue;
                    }

                    $decoded = json_decode($payload, true);
                    // One malformed frame should not discard everything already
                    // streamed.
                    $onEvent(is_array($decoded) ? $decoded : ['raw' => $payload]);
                }

                return $length;
            },
        ]);

        if ($body !== null) {
            curl_setopt($handle, CURLOPT_POSTFIELDS, json_encode($body, JSON_THROW_ON_ERROR));
        }

        curl_exec($handle);
        $status = (int) curl_getinfo($handle, CURLINFO_RESPONSE_CODE);
        $error = curl_error($handle);
        curl_close($handle);

        if ($status >= 400) {
            throw new ApiError($status, $error !== '' ? $error : 'stream failed');
        }
    }
}

File: rust/src/core/sse.rs
//! Server-Sent Events client for the MisarMail streaming endpoints.
//!
//! Both streams frame events as `data: <json>` and close with the sentinel
//! `data: [DONE]`. One of the two is a POST, so the stream is driven off a
//! normal request with a chunked body reader.

use futures_util::StreamExt;
use serde_json::Value;
use tokio::sync::mpsc;

use crate::core::transport::CoreClient;
use crate::errors::MisarMailError;

const DONE: &str = "[DONE]";

/// One decoded SSE frame. `Raw` carries frames that were not valid JSON, so a
/// single malformed frame does not discard everything already streamed.
#[derive(Debug, Clone)]
pub enum SseEvent {
    Data(Value),
    Raw(String),
}

/// Open an SSE endpoint and push each decoded frame onto a channel.
///
/// A channel rather than an `impl Stream` keeps the signature free of the
/// pinning dance callers would otherwise have to perform, and lets the reader
/// drop out simply by dropping the receiver.
pub async fn stream_sse(
    client: &CoreClient,
    method: &str,
    path: &str,
    body: Option<&Value>,
) -> Result<mpsc::Receiver<SseEvent>, MisarMailError> {
    let url = format!("{}{}", client.base_url(), path);
    let verb = reqwest::Method::from_bytes(method.as_bytes()).map_err(|_| MisarMailError::Api {
        status: 0,
        message: format!("unsupported HTTP method: {method}"),
        error_type: "client_error".into(),
    })?;

    let mut request = client
        .http
        .request(verb, &url)
        .bearer_auth(client.api_key())
        .header("Accept", "text/event-stream")
        // A stream has no useful deadline, so it must not inherit the client's
        // per-request timeout.
        .timeout(std::time::Duration::from_secs(3600));

    if let Some(payload) = body {
        request = request.json(payload);
    }

    let response = request
        .send()
        .await
        .map_err(|e| MisarMailError::Network(e.to_string()))?;

    let status = response.status().as_u16();
    if status >= 400 {
        // Errors arrive as a normal JSON body, not as an SSE frame.
        let text = response.text().await.unwrap_or_default();
        let data: Value = serde_json::from_str(&text).unwrap_or(Value::Null);
        return Err(MisarMailError::Api {
            status,
            message: data
                .get("error")
                .and_then(Value::as_str)
                .unwrap_or("stream failed")
                .to_string(),
            error_type: "api_error".into(),
        });
    }

    let (tx, rx) = mpsc::channel(64);

    tokio::spawn(async move {
        let mut bytes = response.bytes_stream();
        let mut buffer = String::new();

        while let Some(chunk) = bytes.next().await {
            let Ok(chunk) = chunk else { break };
            buffer.push_str(&String::from_utf8_lossy(&chunk));

            while let Some(index) = buffer.find('\n') {
                let line: String = buffer.drain(..=index).collect();
                let line = line.trim_end();

                let Some(payload) = line.strip_prefix("data:") else {
                    continue;
                };
                let payload = payload.trim();

                if payload == DONE {
                    return;
                }
                if payload.is_empty() {
                    continue;
                }

                let event = match serde_json::from_str::<Value>(payload) {
                    Ok(value) => SseEvent::Data(value),
                    Err(_) => SseEvent::Raw(payload.to_string()),
                };
                if tx.send(event).await.is_err() {
                    return; // receiver dropped — stop reading
                }
            }
        }
    });

    Ok(rx)
}
